### PR TITLE
IAT-3256 - Validate Braille Data

### PR DESCRIPTION
**Pull request by alex-ponce**
Ported Braille data validation from C# version

### CODE REVIEW

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/TabulatorUtil.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package org.opentestsystem.ap.item.validator.tabulator;
+
+import org.opentestsystem.ap.item.validator.model.ItemTabulationData;
+import org.opentestsystem.ap.item.validator.model.TabulationError;
+import org.opentestsystem.ap.item.validator.model.ValidationError;
+
+public class TabulatorUtil
+{
+
+    public static ValidationError createValidationError(TabulationError tabulationError,
```
Comment (millerts) [inline ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/TabulatorUtil.java]: Instead of a util can you add this to the ValidationError class?

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+ * Validates braille files and their associated item configuration
+ */
+@Component
+public class BrailleTypeTabulator implements Tabulator {
+    @Override
+    public Collection<ValidationError> validate(String itemId, ItemTabulationData itemTabulationData) {
+        List<AttachmentData> attachmentList = new ArrayList<>();
+        // Retrieve attachments from Item or Passage
+        // Map to generic AttachmentData model
+        if (itemTabulationData.getItem().isAssessment()) {
```
Comment (millerts) [inline ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Can we break this out into a separate private method called something like `List<Attachment> findItemAttachments(Item item)`?  It'll make it easier to read the validation logic.

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+        // validationErrors contains all errors generated by attachments
+        Set<ValidationError> validationErrors = new HashSet<>();
+        // includedSubTypes is a list of all attachment subtype values for this item
+        List<String> includedSubTypes = new ArrayList<>();
+        // includedTranscriptSubTypes is a list of all 'transcript' subtype values for this item
+        List<String> includedTranscriptSubTypes = new ArrayList<>();
+
+        attachmentList.forEach(attachment -> {
+            // Check attachment is braille
+            if (isValidBrailleType(attachment.getType())) {
```
Comment (millerts) [inline ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Rather than a for each I think you should have:

```
attachmentList.filter(isValidBrailleType(attachment.getType)).forEach();
```

It would remove one level of nesting you have within the for each.

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+                    String brailleSupport = getBrailleSupport(includedSubTypes);
+                    if (brailleSupport.equals(BRAILLE_SUPPORT_UNEXPECTED)) {
+                        validationErrors.add(
+                            TabulatorUtil.createValidationError(TabulationError.BRAILLE_FILE_SET_UNEXPECTED_PATTERN,
+                                String.format("brailleTypes=%s", String.join(",", includedSubTypes)),
+                                itemTabulationData));
+                    }
+                    // Retrieve BrailleType from metadata
+                    String brailleTypeMeta = itemTabulationData.getMetadata().getBrailleType();
+                    if (brailleTypeMeta.equals("Not Braillable")) {
```
Comment (millerts) [inline ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: This is a bit hard to parse for me without the context.  Basically, a braille file can only have one of these errors?

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+            // Check attachment is braille
+            if (isValidBrailleType(attachment.getType())) {
+                // Check if file attribute is populated
+                if (attachment.getFile().isEmpty()) {
+                    validationErrors
+                        .add(TabulatorUtil.createValidationError(TabulationError.ATTACHMENT_MISSING_FILE_ATTRIB,
+                            String.format("attachType=%s", attachment.getType()),
+                            itemTabulationData));
+                } else {
+                    // Validate file name
```
Comment (millerts) [inline ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Rather than do an if/else if you do a `return;` on line 103 then it would short circuit the code removing one level of nesting.

What about pulling this large else conditional block into its own private method with a clear name? It'd make it easier for readability in the main flow.

File: ap-item-validation/src/test/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulatorTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    public TemporaryFolder folder = new TemporaryFolder();
+    private BrailleTypeTabulator tabulator;
+
+    @Before
+    public void setUp() {
+        tabulator = new BrailleTypeTabulator();
+    }
+
+    @Test
+    public void shouldHaveExpectedItemErrors() {
```
Comment (millerts) [inline ap-item-validation/src/test/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulatorTest.java]: Does this method test all the permutations of the tabulator logic you wrote?  If not, I recommend renaming this to better indicate what is being tested.

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/TabulatorUtil.java (hunk 1 above)
Comment (alex-ponce) [reply ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/TabulatorUtil.java]: Done

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java (hunk 2 above)
Comment (alex-ponce) [reply ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Done

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java (hunk 3 above)
Comment (alex-ponce) [reply ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Done

File: ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java (hunk 5 above)
Comment (alex-ponce) [reply ap-item-validation/src/main/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulator.java]: Done

File: ap-item-validation/src/test/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulatorTest.java (hunk 6 above)
Comment (alex-ponce) [reply ap-item-validation/src/test/java/org/opentestsystem/ap/item/validator/tabulator/metadata/BrailleTypeTabulatorTest.java]: I added one test for item, one for passage and one for "transcript" braille files. This does not include all permutations of validation logic. All variations of braille file names are not known.
